keps/sig-cli/3515-kubectl-explain-openapiv3/README.md

```diff
@@ -0,0 +1,958 @@
+wiping nullable fields.
+
+### Goals
+
+1. Provide the new richer type information specified by OpenAPI v3 within kubectl explain
+2. Have a more maintainable `text/template` based approach to printing
+3. Fallback to old `explain` implementation if cluster does not expose OpenAPI v3 data.
+4. Provide multiple new output formats for kubectl explain:
+    * human-readable plaintext
+    * markdown
+    * maybe others
+5. (Optional?) Allow users to specify their own templates for use with kubectl
+  explain (there may be interesting use cases for this)
+
+### Non-Goals
+
+1. "Fix" openapi v3 to openapi v2 conversion
+  This is a non-goal for two reasons:
+    * These formats are not compatible, and there WILL be data loss and inaccuracy
+    * This negates the benefits of using OpenAPI v3 for the richer type information
+2. Provide general-purpose OpenAPI visualization.
+
+
+## Proposal
+
+
+### Basic Usage
+
+The following user experience should be possible with `kubectl explain`
+
+```shell
+kubectl explain pods.spec
+```
+
+Output should be familiar to users of today's `kubectl explain`, except new
+information from the OpenAPI v3 spec is now populated.
+
+Note: Feature during development will be gated by an experimental flag. The commands
+shown here elide the experimental flag for clarity.
+
+### Built-in Template Options
+#### Plaintext
+
+
+```shell
+kubectl explain pods
+```
+or
+
+```shell
+kubectl explain pods --output plaintext
+```
+
+The plaintext output format is the default and should be crafted to be as close
+as the existing `explain` output in use before this KEP.
+
+#### OpenAPIV3 (raw json)
+
+```shell
+kubectl explain pods --output openapiv3
+```
+
+To get raw OpenAPI v3 data for a certain resource today involves:
+1.) setting up kubectl proxy
+2.) fetching the correct path at `/openapi/v3/<group>/<version>`
+3.) filtering out unwanted results
+
+This command is useful not only for its convenience, but also other visualizations
+may be built upon the raw output if we opt not to support a first-class custom
+template solution in the future.
+
+#### HTML
+
+```shell
+kubectl explain pods --output html
+```
+
+Similarly to [godoc](https://pkg.go.dev), the we suggest to provide a searchable,
+navigable, generated webpage for the kubernetes types of whatever cluster kubectl
+is talking to.
+
+#### Markdown
+
+```shell
+kubectl explain pods --output md
+```
+
+When using the `md` template, a markdown document is printed to stdout, so it
+might be saved and used for a documentation website, for example.
+
+### Risks and Mitigations
+
+#### OpenAPI V3 Not Available
+
+##### Risk
+
+OpenAPI v3 data is not available in the current cluster.
+
+##### Mitigation
+
+###### If the user does not provide an --output argument
+
+This is a "compatibility mode" where users will see OpenAPI v3 plaintext template
+if the data is available, or fallback to old `kubectl explain` behavior for openapi v2.
```

Comment:
I'm taking `--experimental-openapiv3` to mean that the user wants ONLY openapiv3, and if it is unavailable, then an error is shown.

That said, I also question the value of an additional flag `--experimental-openapiv3` when usage of `--output` also communicates the same thing the way I've specified. 

I've only kept it in since it makes it clear the feature is experimental, thoughts would be appreciated.

Reply:
I've removed `--experimental-openapiv3` in favor of `--experimental-output`, which will be renamed to `--output` once GA.